### PR TITLE
allow unlocking onchain privacy groups

### LINKED ISSUE #2693: Flexible Privacy Group can not unlock.
# summary

- This occurs when flexiblePrivacyGroup is used.
- If the lock is acquired with setLockState, it cannot be unlocked.

# Steps to Reproduce

- Launches the besu container, which is available in flexible privacy group mode.
- create a privacy group.
- get the lock with lock function.
- release the lock with unlock function, but it can't release the lock.

## sample script
```
const Web3 = require("web3");
const EEAClient = require("web3js-quorum");

const { tessera, besu } = require("../config/keys.js");

const node1 = new EEAClient(new Web3(besu.node1.url));

const privacyProxyAbi = require("./PrivacyProxy.json").output.abi
// https://github.com/ConsenSys/web3js-quorum/blob/master/solidity/PrivacyProxy.json

module.exports = async () => {
  const onChainPrivacyGroupCreationResult = await node1.eth.flexiblePrivacyGroup.create(
    {
      participants: [tessera.node1.publicKey, tessera.node2.publicKey],
      enclaveKey: tessera.node1.publicKey,
      privateKey: besu.node1.privateKey
    }
  );
  console.log("Created new on-chain privacy group:");
  console.log(onChainPrivacyGroupCreationResult.privacyGroupId);

  console.log("-- canExecute before lock--")
  await canExecute(node1, {
      enclaveKey: tessera.node1.publicKey,
      privacyGroupId: onChainPrivacyGroupCreationResult.privacyGroupId,
      privateKey: besu.node1.privateKey
    }
  )
  
  console.log("-- get lock --")
  await setLockState(node1, {
      enclaveKey: tessera.node1.publicKey,
      privacyGroupId: onChainPrivacyGroupCreationResult.privacyGroupId,
      privateKey: besu.node1.privateKey,
      lock: true
    }
  )

  console.log("-- canExecute after lock--")
  await canExecute(node1, {
      enclaveKey: tessera.node1.publicKey,
      privacyGroupId: onChainPrivacyGroupCreationResult.privacyGroupId,
      privateKey: besu.node1.privateKey
    }
  )

  console.log("-- get unlock --")
  await setLockState(node1, {
      enclaveKey: tessera.node1.publicKey,
      privacyGroupId: onChainPrivacyGroupCreationResult.privacyGroupId,
      privateKey: besu.node1.privateKey,
      lock: false
    }
  )

  console.log("-- canExecute after unlock--")
  await canExecute(node1, {
      enclaveKey: tessera.node1.publicKey,
      privacyGroupId: onChainPrivacyGroupCreationResult.privacyGroupId,
      privateKey: besu.node1.privateKey
    }
  )
}

async function setLockState(web3, options) {
  const contract = new web3.eth.Contract(privacyProxyAbi);
  const functionAbi = contract._jsonInterface.find((e) => {
    return e.name === (options.lock ? "lock" : "unlock");
  });

  const functionCall = {
    to: "0x000000000000000000000000000000000000007c",
    data: functionAbi.signature,
    privateFrom: options.enclaveKey,
    privacyGroupId: options.privacyGroupId,
    privateKey: options.privateKey,
  };

  let transactionHash = await web3.priv.generateAndSendRawTransaction(functionCall)
  let transactionReceipt = await web3.priv.waitForTransactionReceipt(transactionHash);
  console.log("TransactionReceipt.status: ", transactionReceipt.status)
  console.log("TransactionReceipt.output: ", transactionReceipt.output)
}

async function canExecute(web3, options) {
  const contract = new web3.eth.Contract(privacyProxyAbi);
  const functionAbi = contract._jsonInterface.find((e) => {
    return e.name === ("canExecute");
  });

  const functionCall = {
    to: "0x000000000000000000000000000000000000007c",
    data: functionAbi.signature,
    privateFrom: options.enclaveKey,
    privacyGroupId: options.privacyGroupId,
    privateKey: options.privateKey,
  };

  let transactionHash = await web3.priv.generateAndSendRawTransaction(functionCall)
  let transactionReceipt = await web3.priv.waitForTransactionReceipt(transactionHash);
  console.log("TransactionReceipt.status: ", transactionReceipt.status)
  console.log("TransactionReceipt.output: ", transactionReceipt.output)
}

if (require.main === module) {
  module.exports();
}
```

## console log
```
Waiting for transaction to be mined ...
Created new on-chain privacy group:
bHPtl5dGLKn/hOUv5Uq7h7VRSJNPe8IT07Kk1H9kcEI=
-- canExecute before lock--
Waiting for transaction to be mined ...
TransactionReceipt.status:  0x1
TransactionReceipt.output:  0x0000000000000000000000000000000000000000000000000000000000000001
-- get lock --
Waiting for transaction to be mined ...
TransactionReceipt.status:  0x1
TransactionReceipt.output:  0x
-- canExecute after lock--
Waiting for transaction to be mined ...
TransactionReceipt.status:  0x0
TransactionReceipt.output:  0x
-- get unlock --
Waiting for transaction to be mined ...
TransactionReceipt.status:  0x0
TransactionReceipt.output:  0x
-- canExecute after unlock--
Waiting for transaction to be mined ...
TransactionReceipt.status:  0x0
TransactionReceipt.output:  0x
```

# Versions 
- BESU: besu/v21.7.2/linux-x86_64/adoptopenjdk-java-11
- web3js-quorum: 21.7.0-rc1

### PR DESCRIPTION
**Pull request by taccatisid**
## PR description

Prior to this commit, the only allowed transactions in locked onchain privacy groups were those making the `addParticipants` call to *any* contract--not just the proxy. This commit ensures that:
- while locked, only the group management proxy can be called
- while locked, all functions of the management proxy can be called

Corollary, this commit allows unlocking a group using `unlock` instead of implicitly unlocking it via `addParticipants`. This fixes #2693.

## Fixed Issue(s)

fixes #2693

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java [hunk 1]
```diff
@@ -315,6 +310,26 @@ private String getRemovedParticipantFromParameter(final Bytes input) {
     return input.slice(4).toBase64String();
   }
 
+  public boolean isTargettingOnchainPrivacyProxy(final PrivateTransaction privateTransaction) {
+    return privateTransaction.getTo().isPresent()
+        && privateTransaction.getTo().get().equals(Address.ONCHAIN_PRIVACY_PROXY);
+  }
+
+  private boolean isAddingParticipant(final PrivateTransaction privateTransaction) {
+    return privateTransaction
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java:319]: should that not check for the proxy address as well?

**Review by macfarla — COMMENTED** [review]
is there a test for the new behaviour?

**Comment by taccatisid** [comment]
> is there a test for the new behaviour?

I added an acceptance test that locks & unlocks and tries to call a contract while [un]locked.

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/transaction/PrivacyTransactions.java [hunk 2]
```diff
@@ -78,6 +78,11 @@ public LockOnChainPrivacyGroupTransaction privxLockPrivacyGroupAndCheck(
     return new LockOnChainPrivacyGroupTransaction(privacyGroupId, locker, signer);
   }
 
+  public UnlockOnChainPrivacyGroupTransaction privxUnlockPrivacyGroupAndCheck(
```
[the remark this answers is not in the document]
Comment (taccatisid) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/transaction/PrivacyTransactions.java:81]: I copied the camel cased `OnChain` from the lock function above. Overall we have more OnChain that Onchain:
```
besu $ rg OnChain -g '*.java' | wc -l
347
besu $ rg Onchain -g '*.java' | wc -l
123
```
But I do agree that treating onchain as one word is nicer.

I don't want to mix both casing in one file and changing the existing function trickles up the call stack. Would it make sense to do a separate PR to rename all OnChains in besu to Onchains?

**Review by pinges — APPROVED** [review]
LGTM

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/transaction/PrivacyTransactions.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (taccatisid) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/transaction/PrivacyTransactions.java:81]: Moved into separate issue https://github.com/hyperledger/besu/issues/2750

**Pull request merged**